### PR TITLE
VQ-VAE for CSIRO HipMRI dataset

### PR DESCRIPTION
**Pull request by kirra-4**
This pull request introduces a VQ-VAE generative model for HipMRI images. The model trains on the CSIRO HipMRI Prostate Cancer Dataset and produces reconstructed images with an SSIM greater than 0.6, averaging approximately 0.8. 

Files Added include:

dataset.py : loading train, test, and validation dataloaders from file paths containing Nifti files
modules.py : contains all torch models that contribute to the VQ-VAE, as well as the entire VQ-VAE model. This includes the Encoder and Decoder, which utilise Residual neural network stacks, and the Vector Quantiser. 
train.py : contains model set up, model hyperparameters, model training loop with metric tracking. Is executable and will produce model reconstruction image plots, and metric plots. 
predict.py : contains image generation plots to compare reconstruction against original images. 

The model achieves an SSIM score of 0.84 on validation set and 0.85 on test set, training in 57 minutes for 150 epochs. For more details of outputs and model parameters, see README.md.

Thank you for reviewing this repository, 
Kirra Fletcher s4745168

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem: (4.5 Marks)
Very clear reconstructed images, with an SSIM over 0.8;
Inappropriate usage of validation and test sets (see below) -0.5

2. Algorithm demonstrated and functions: (3 Marks)

4. Design: (1 Mark)
Modularized design, good practice using util script

5. Commenting:
Sufficient docstrings and comments (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:

- The validation set seems to be sampled from the training set, this is fine given your split, but all training, validation, test sets are available on Rangpur;
- You used the validation set in the same way as you used the test set, while ideally the validation should be included within the training process to compare the model's performance on seen and unseen data, to see if it overfits and for any hyperparameter tuning;
- Test should be done in `predict.py` for modularization purpose;
- Hyperparameters are given, but no justification on the tuning process;
- No reference.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>Data leakage found. Inappropriate usage of validation and test sets.</td><td>-1</td></tr>
	<tr><td></td><td>Difficulty : Hard. VQVAE (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. No references.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by kirra-4** [comment]
I have references to the paper and the dataset in hyperlinks in the first part of my ReadMe, are there other references I am missing?

**Comment by shakes76** [comment]
Good work. Cant merge because of conflicting changes to main repo files. Please update for merge, doesn't affect grade.
